### PR TITLE
New make:migration command

### PR DESCRIPTION
**Pull request by weaverryan**
This replaces #10.

There are 2 cases:

(A) When there ARE schema changes:

screen shot 2017-12-29 at 11 07 47 pm

(B) When there are NOT schema changes:

screen shot 2017-12-29 at 11 07 14 pm

This includes a BC break on `MakerInterface`, but it will only affect people who have created custom makers (and they'll get a clear error).

### CODE REVIEW

File: src/Command/MakerCommand.php [hunk 1]
```diff
@@ -79,7 +82,7 @@ protected function interact(InputInterface $input, OutputInterface $output)
                 continue;
             }
 
-            $value = $this->io->ask($argument->getDescription(), $argument->getDefault(), [Validator::class, 'notBlank']);
+            $value = $this->io->ask($argument->getDescription(), $argument->getDefault(), array(Validator::class, 'notBlank'));
```
Comment (javiereguiluz) [inline src/Command/MakerCommand.php]: Please, revert this change. We don't use the legacy `array()` notation in this bundle.

File: src/ExtraGenerationMakerInterface.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ * Interface to do extra code generation in a maker.
+ *
+ * @author Ryan Weaver <ryan@knpuniversity.com>
+ */
+interface ExtraGenerationMakerInterface
+{
+    /**
+     * Called after normal code generation.
+     *
+     * @param ConsoleStyle $io
```
Comment (javiereguiluz) [inline src/ExtraGenerationMakerInterface.php]: The PHPdoc is not needed because we use type-hints.

File: src/Maker/MakeMigration.php [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    public function setApplication(Application $application)
+    {
+        $this->application = $application;
+    }
+
+    public function configureCommand(Command $command, InputConfiguration $inputConf)
+    {
+        $command
+            ->setDescription('Creates a new migration based on database changes.')
+            ->addOption('db', null, InputOption::VALUE_REQUIRED, 'The database connection to use for this command.')
```
Comment (javiereguiluz) [inline src/Maker/MakeMigration.php]: I'd remove *"...to use for this command"* in all descriptions because it's redundant. Also, I'd like better descriptions. For example, this doesn't help me -> `db = The database connection.` I can't provide a database connection to a console command. Is this the name of the database connection? The name of the Doctrine database connection? etc.

Comment (javiereguiluz) [inline src/Maker/MakeMigration.php]: Should all these options define default values for the common case where there's only 1 database, 1 em, etc.?

File: src/Maker/MakeMigration.php [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                'db' => $input->getOption('db'),
+                'em' => $input->getOption('em'),
+                'shard' => $input->getOption('shard'),
+            ),
+        );
+    }
+
+    public function getFiles(array $params): array
+    {
+        return array();
```
Comment (javiereguiluz) [inline src/Maker/MakeMigration.php]: More `array()` notation to replace by `[]` Thanks!

File: src/Maker/MakeMigration.php [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        $options['command'] = 'doctrine:migrations:diff';
+        $generateMigrationCommand = $this->application->find('doctrine:migrations:diff');
+        $generateMigrationCommand->run(new ArgvInput($options), $io);
+    }
+
+    public function writeNextStepsMessage(array $params, ConsoleStyle $io)
+    {
+        $io->text(array(
+            'Next: Edit Version(lastVersion) to customize the migration.',
+            'Find the documentation at <fg=yellow>https://symfony.com/doc/current/bundles/DoctrineMigrationsBundle/index.html</>',
```
Comment (javiereguiluz) [inline src/Maker/MakeMigration.php]: Could we standardize these doc links in all commands? `Find the documentation at...` looks too long to me. What about `See ...` ?

File: src/Command/MakerCommand.php (hunk 1 above)
Comment (Pierstoval) [reply src/Command/MakerCommand.php]: And it's available since PHP 5.4. Anyone using this bundle will probably have at least 5.5 or 5.6

File: src/Maker/MakeMigration.php (hunk 3 above)
Comment (Pierstoval) [reply src/Maker/MakeMigration.php]: I think you can already get the default connection from Doctrine, as well as for the entity manager, especially if you have only one connection

File: src/Maker/MakeMigration.php [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+            ->addOption('db', null, InputOption::VALUE_REQUIRED, 'The database connection to use for this command.')
+            ->addOption('em', null, InputOption::VALUE_OPTIONAL, 'The entity manager to use for this command.')
+            ->addOption('shard', null, InputOption::VALUE_REQUIRED, 'The shard connection to use for this command.')
+            ->setHelp(file_get_contents(__DIR__.'/../Resources/help/MakeMigration.txt'))
+        ;
+    }
+
+    public function interact(InputInterface $input, ConsoleStyle $io, Command $command)
+    {
+    }
```
Comment (Pierstoval) [inline src/Maker/MakeMigration.php]: Why implement interact when it's empty?

File: tests/Command/FunctionalTest.php [hunk 7]
```diff
@@ -207,6 +208,12 @@ public function getCommandTests()
                 'AppCustomAuthenticator',
             ),
         );
+
+        yield 'migration' => array(
+            new MakeMigration(),
+            array(
+            ),
+        );
```
Comment (Pierstoval) [inline tests/Command/FunctionalTest.php]: Again please use short-syntax for arrays 👍

**Comment by weaverryan** [comment]
This is ready to be reviewed :) - screenshots back in the original description. This also uses the new test system, which means there are real functional tests that actually run the commands and assert that the 2 outputs above occur.

File: src/Maker/MakeMigration.php [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+    public function writeSuccessMessage(array $params, ConsoleStyle $io)
+    {
+        if (false !== strpos($this->migrationOutput, 'No changes detected')) {
+            $io->warning([
+                'No database changes were detected.'
+            ]);
+            $io->text([
+                'The database schema was compared with your mapping information and they are already in sync: there is no generation to migrate.',
```
Comment (fabpot) [inline src/Maker/MakeMigration.php]: I don't understand the "there is no generation to migrate." part :) I suppose it should be "there is no migration to generate."

Comment (weaverryan) [reply src/Maker/MakeMigration.php]: Whoooops :)

**Comment by weaverryan** [comment]
Comment addressed (and a few other things). Tests are passing, ready for a final review :).

And sorry for the noise in the diff: this command is a bit different, so we needed to expand the maker system a bit and also the tests, which affected some unrelated things.

File: src/Maker/MakeMigration.php [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+        $generateMigrationCommand = $this->application->find('doctrine:migrations:diff');
+
+        $commandOutput = new BufferedOutput($io->getVerbosity());
+        $generateMigrationCommand->run(new ArgvInput($options), $commandOutput);
+        $this->migrationOutput = $commandOutput->fetch();
+    }
+
+    public function writeSuccessMessage(array $params, ConsoleStyle $io)
+    {
+        if (false !== strpos($this->migrationOutput, 'No changes detected')) {
```
Comment (javiereguiluz) [inline src/Maker/MakeMigration.php]: I don't like this logic because it looks fragile ... but there's no other possible way to detect if a migration was generated or not. If MigrationsBundle supported JSON as one of its output formats ... but it doesn't.

File: src/Maker/MakeMigration.php [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+            $io->warning([
+                'No database changes were detected.',
```
Comment (javiereguiluz) [inline src/Maker/MakeMigration.php]: Just wondering if we could reword the messages in case no migration is needed:

Before:

```
[WARNING] No database changes were detected.

The database schema was compared with your mapping information and they are
already in sync: no migration was generated.
```

After:

```
[WARNING] No migration was generated.

The database schema and the application mapping information are already in sync.
```

File: src/Maker/MakeMigration.php (hunk 9 above)
Comment (weaverryan) [reply src/Maker/MakeMigration.php]: You're exactly right. So I'm depending on our tests to catch this :/
